roles/jd-server/src/lib/job_declarator/mod.rs

```diff
@@ -65,23 +88,115 @@ impl JobDeclaratorDownstream {
+                mempool::error::JdsMempoolError::NoClient,
+            ))?;
+        for txid in transactions_to_be_retrieved {
+            let transaction = client
+                .get_raw_transaction(&txid.to_string(), None)
+                .await
+                .map_err(|e| JdsError::MempoolError(mempool::error::JdsMempoolError::Rpc(e)))?;
+            let txid = transaction.txid();
+            mempool::JDsMempool::add_tx_data_to_mempool(
+                mempool_.clone(),
+                txid,
+                Some(transaction.clone()),
+            );
+            new_transactions.push(transaction);
+        }
+        for transaction in new_transactions {
+            self_mutex
+                .clone()
+                .safe_lock(|a| {
+                    for transaction_with_state in &mut a.declared_mining_job.1 {
+                        match transaction_with_state {
+                            TransactionState::Present(_) => continue,
+                            TransactionState::ToBeRetrievedFromNodeMempool(_) => {
+                                *transaction_with_state =
+                                    TransactionState::Present(transaction.txid());
+                                break;
+                            }
+                            TransactionState::Missing => continue,
+                        }
+                    }
+                })
+                .map_err(|e| JdsError::PoisonLock(e.to_string()))?
+        }
+        Ok(())
+    }
+
+    fn get_block_hex(
+        self_mutex: Arc<Mutex<Self>>,
+        message: SubmitSolutionJd,
+    ) -> Result<String, Box<JdsError>> {
+        let (last_declare_, transactions_with_state, _) = self_mutex
+            .safe_lock(|x| x.declared_mining_job.clone())
+            .map_err(|e| JdsError::PoisonLock(e.to_string()))?;
+        let mempool_ = self_mutex
+            .safe_lock(|x| x.mempool.clone())
+            .map_err(|e| JdsError::PoisonLock(e.to_string()))?;
+        let last_declare = last_declare_.ok_or(JdsError::NoLastDeclaredJob)?;
+        let mut transactions_list: Vec<Transaction> = Vec::new();
+        for tx_with_state in transactions_with_state.iter().enumerate() {
+            if let TransactionState::Present(txid) = tx_with_state.1 {
+                let tx_ = match mempool_.safe_lock(|x| x.mempool.get(txid).cloned()) {
+                    Ok(tx) => tx,
+                    Err(e) => return Err(Box::new(JdsError::PoisonLock(e.to_string()))),
+                };
+                let tx = tx_.ok_or(JdsError::ImpossibleToReconstructBlock(
+                    "Missing transactions".to_string(),
+                ));
+                if let Ok(Some(tx)) = tx {
+                    transactions_list.push(tx);
+                } else {
+                    return Err(Box::new(JdsError::ImpossibleToReconstructBlock(
+                        "Missing transactions".to_string(),
+                    )));
+                }
+            } else {
```

Comment:
you are assuming that when you receive the solution the task that fetch tx data from the bitcoin node has finished. This assumption is not always true. You can put an ref count with an atomic bool set to false in that tuple:
```rust
....
    declared_mining_job: (
        Option<DeclareMiningJob<'static>>,
        Vec<TransactionState>,
        Vec<u16>,
    ),
```
will be:
```rust
    declared_mining_job: (
        Option<DeclareMiningJob<'static>>,
        Vec<TransactionState>,
        Vec<u16>,
        Arc<AtomicBool>,
    ),
```
And you set it to true when retrieve_transactions_via_rpc finish. If for now you don't want to handle it cause is very unlikely I'm not sure that the right thing to do is to close the connection with downstream.

Reply:
The assumption is correct but just cause we don't create a task and we await the future before continue with the main loop (that invalidate the all point of this PR) so we should have a parallel task that fetch txs, and when we have it the assumption is not anymore true, so my comment above still need to be addressed.  

https://github.com/stratum-mining/stratum/pull/772#discussion_r1527129448

Reply:
@lorbax is working on this, he will push a commit on the PR he opened against my branch in order to check it and properly manage this case